### DISCUSSION TITLE
Setting default shell (or other step metadata) conditionally in workflows

### DISCUSSION THREAD

**Discussion by ferdnyc** [community/community#27143] [Actions]
<p>I just posted this to the <a href="https://github.com/msys2/setup-msys2/issues/102" rel="noopener nofollow ugc">setup-msys2 issue tracker</a>, but I thought I’d bounce it off the folks here as well:</p>
<p>So, I’m kind of banging my head against a wall with something, and I wanted to see if anyone has any clever solutions.</p>
<p>I have a project which builds in CMake, and the <code>cmake</code> commands are actually all <strong>the same</strong> whether running on Linux, macOS, or Windows. I’ve worked very hard to keep things that way, in fact, so that the build tooling could be as platform-agnostic as possible.</p>
<p>So, I have a workflow written with build steps that are largely shared between all platforms in the matrix. (Each OS has its own dependency step, gated with an <code>if: ${{ runner.os == 'Linux' }}</code> (or <code>'Windows'</code> or <code>'macos'</code>), but that’s the only non-shared config.)</p>
<p>To differentiate the <code>-G</code> argument to <code>cmake</code>, which selects the generator to use for the build system, I’m using <code>action-cond</code> from <a class="mention" href="/haya14busa">@haya14busa</a> which works great:</p>
<pre><code class="lang-auto">     - name: Select CMake generator
        uses: haya14busa/action-cond@v1
        id: generator
        with:
          cond: ${{ runner.os == 'Windows' }}
          if_true: 'MinGW Makefiles'
          if_false: 'Unix Makefiles'
      - name: Build
        run: |
          cmake -B build -S . -G "${{ steps.generator.outputs.value }}" ...
</code></pre>
<p>The problem is, I can’t think of <strong>any</strong> way to do the same thing for the MSYS2 shell. If I wanted to take advantage of the “Default shell” option <a href="https://github.com/msys2/setup-msys2#default-shell" rel="noopener nofollow ugc">from the README</a>, I’d end up setting it as the default for <em>every</em> shell, including the ones that run on Linux and macOS, which obviously won’t work! I’ve tried 100 different ways to conditionalize the setting, and they all failed:</p>
<ol>
<li>Add an <code>if:</code> to the <code>defaults:</code> section of the config: Syntax error</li>
<li>Add another <code>haya14busa/action-cond@v1</code> step that sets either <code>msys2 {0}</code> or <code>bash</code>, and use <code>shell: ${{ steps.select_shell.outputs.value }}</code> in subsequent steps: Syntax error, apparently the <code>steps</code> object is not accessible from the metaconfiguration of subsequent steps.</li>
<li>Capture the output into an environment variable, the same way I can with <code>CC: ${{ matrix.compiler }}</code>:<pre><code class="lang-auto">env:
  RUNSHELL: ${{ steps.select_shell.outputs.value }}
steps:
   - name: Select shell...
     id: select_shell
     with:
      cond: ${{ runner.os == 'Windows' }}
      if_true: 'msys2 -c'
      if_false: 'bash -c'
     ...
  - name: Build
    run: |
      $RUNSHELL 'cmake...'
</code></pre>
(In my defense, I never expected that one to work, and it doesn’t. Same issue with accessing <code>steps</code> before it’s defined.)</li>
<li>The same thing, but set the <code>env</code> in every step that needs it:<pre><code class="lang-auto">steps:
   - name: Select shell...
     id: select_shell
     with:
      cond: ${{ runner.os == 'Windows' }}
      if_true: 'msys2 -c'
      if_false: 'bash -c'
     ...
  - name: Build
    env:
      RUNSHELL: ${{ steps.select_shell.outputs.value }}
    run: |
      $RUNSHELL 'cmake...'
</code></pre>
That actually came close to working, believe it or not! Actually worked on Linux and macOS. Blew up in the default Powershell on Windows, though.</li>
<li>Set up a completely separate <strong>job</strong> config before the build job, with the same matrix, and have it relay the results of <code>action-cond</code> into an <code>outputs</code>. Have the second job <code>needs: </code> the first, and set <code>shell: ${{ needs.job1.outputs.shell }}</code> on each step: Syntax error, it seems the <code>needs</code> context isn’t any more accessible for defining steps than the <code>steps</code> context is.</li>
</ol>
<p>At this point I feel like I’ve exhausted all possibilities. Any suggestions, or do I really have to take my nice, cross-platform workflows and duplicate all of the run steps, just so I can get Windows to run them using the correct shell? :disappointed:</p>

**Comment by ferdnyc** [comment, marked as answer]
<p>Over at my setup-msys2 issue, we did some brainstorming and finally came up with this, which works unexpectedly well.</p>
<p>It uses what appears to be undocumented syntax for the <code>matrix</code> context in the Actions yaml, and its contents will get flagged as containing syntax errors by Github’s web workflow editor. Nevertheless, it works exactly as you’d hope it would. All credit to <a class="mention" href="/eine">@eine</a> for coming up with the necessary matrix config wizardry.</p>
<pre><code class="lang-auto">jobs:
  build:
    runs-on: ${{ matrix.sys.os }}
    strategy:
      matrix:
        sys:
          - { os: windows-latest, shell: 'msys2 {0}' }
          - { os: ubuntu-latest,  shell: bash  }
          - { os: macos-latest,   shell: bash  }
        compiler:
          - { cc: gcc,   cxx: g++ }
          - { cc: clang, cxx: clang++ }
    defaults:
      run:
        shell: ${{ matrix.sys.shell }}
    env:
      CC: ${{ matrix.compiler.cc }}
      CXX: ${{ matrix.compiler.cxx }}
</code></pre>

**Reply by dabrahams** [reply]
I'm not a YAML or actions expert but to me this doesn't look like an "undocumented syntax," so much as a technique that isn't shown explicitly in the documentation.  It _looks_ to me that there's nothing indicating that matrix property array elements can't themselves be arbitrary YAML objects.  Am I missing anything?

Also, while your technique might be a little easier to understand, It looks like [`jobs.<job_id>.strategy.matrix.include`](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations) can be used to achieve the same things.  What do you think?

**Reply by ferdnyc** [reply]
This is a >2 year old question that was covered pretty extensively in the discussion at that setup-msys2 issue I linked to initially; I'm not particularly inclined to re-hash it all here. But, briefly...

> It _looks_ to me that there's nothing indicating that matrix property array elements can't themselves be arbitrary YAML objects.

There's a _lot_ that implies it. Not least, the fact that **only** this syntax works.

In theory, these should be equivalent, and one or both of them should also work:

```yaml
strategy:
  matrix:
    sys:
      - os: windows-latest
        shell: 'msys2 {0}
      - os: ubuntu-latest
        shell: bash
      - os: macos-latest
        shell: bash
   compiler:
     - cc: gcc
       cxx: g++
     - cc: clang
       cxx: clang++
```

```yaml
strategy:
  matrix:
    sys:
      - 
        os: windows-latest
        shell: 'msys2 {0}
      - 
        os: ubuntu-latest
        shell: bash
      - 
        os: macos-latest
        shell: bash
   compiler:
     - 
       cc: gcc
       cxx: g++
     - 
       cc: clang
       cxx: clang++
```

...But they don't. Actions will reject either syntax as a workflow error. If this were truly a supported, intentional feature, it wouldn't matter how you got to the data structure as long as you got there.

**Reply by dabrahams** [reply]
Ah well I didn't have that information before. I was just going by GitHub's documentation, which places no restrictions on these objects, and now I can report it as a bug that it doesn't support general YAML. Anyway, sorry, I didn't mean to irritate you by reopening the question; I'm just trying to better understand this black box and poking at this discussion has been fruitful. Did you think about using the `matrix.include` feature for this?

**Reply by ferdnyc** [reply]
https://github.com/msys2/setup-msys2/issues/102#issuecomment-756963300

https://github.com/msys2/setup-msys2/issues/102#issuecomment-757387822

**Reply by dabrahams** [reply]
I see... I think.
